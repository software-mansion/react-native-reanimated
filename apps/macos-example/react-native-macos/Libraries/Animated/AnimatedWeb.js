/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 */

'use strict';

import AnimatedImplementation from './AnimatedImplementation';

export default {
  ...AnimatedImplementation,
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  div: (AnimatedImplementation.createAnimatedComponent('div'): $FlowFixMe),
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  span: (AnimatedImplementation.createAnimatedComponent('span'): $FlowFixMe),
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  img: (AnimatedImplementation.createAnimatedComponent('img'): $FlowFixMe),
};
