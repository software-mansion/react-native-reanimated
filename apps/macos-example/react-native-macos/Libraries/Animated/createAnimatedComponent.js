/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

import View from '../Components/View/View';
import useMergeRefs from '../Utilities/useMergeRefs';
import useAnimatedProps from './useAnimatedProps';
import * as React from 'react';

export type AnimatedComponentType<
  -Props: {+[string]: mixed, ...},
  +Instance = mixed,
> = React.AbstractComponent<
  $ObjMap<
    Props &
      $ReadOnly<{
        passthroughAnimatedPropExplicitValues?: React.ElementConfig<
          typeof View,
        >,
      }>,
    () => any,
  >,
  Instance,
>;

export default function createAnimatedComponent<TProps: {...}, TInstance>(
  Component: React.AbstractComponent<TProps, TInstance>,
): AnimatedComponentType<TProps, TInstance> {
  return React.forwardRef((props, forwardedRef) => {
    const [reducedProps, callbackRef] = useAnimatedProps<TProps, TInstance>(
      // $FlowFixMe[incompatible-call]
      props,
    );
    // $FlowFixMe[incompatible-call]
    const ref = useMergeRefs<TInstance | null>(callbackRef, forwardedRef);

    // Some components require explicit passthrough values for animation
    // to work properly. For example, if an animated component is
    // transformed and Pressable, onPress will not work after transform
    // without these passthrough values.
    // $FlowFixMe[prop-missing]
    const {passthroughAnimatedPropExplicitValues, style} = reducedProps;
    const {style: passthroughStyle, ...passthroughProps} =
      passthroughAnimatedPropExplicitValues ?? {};
    const mergedStyle = {...style, ...passthroughStyle};

    return (
      <Component
        {...reducedProps}
        {...passthroughProps}
        style={mergedStyle}
        ref={ref}
      />
    );
  });
}
