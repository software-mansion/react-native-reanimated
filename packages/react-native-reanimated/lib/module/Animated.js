'use strict';

export { ReanimatedFlatList as FlatList } from './component/FlatList';
export { AnimatedImage as Image } from './component/Image';
export { AnimatedScrollView as ScrollView } from './component/ScrollView';
export { AnimatedText as Text } from './component/Text';
export { AnimatedView as View } from './component/View';
export { addWhitelistedNativeProps, addWhitelistedUIProps } from './ConfigHelper';
export { createAnimatedComponent } from './createAnimatedComponent';
//# sourceMappingURL=Animated.js.map