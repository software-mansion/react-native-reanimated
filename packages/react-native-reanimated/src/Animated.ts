'use strict';
import type { Extrapolate as _Extrapolate } from './interpolateColor';
import type {
  SharedValue as _SharedValue,
  AnimatedTransform as _AnimatedTransform,
  AnimateStyle as _AnimateStyle,
  StylesOrDefault as _StylesOrDefault,
  EasingFunction as _EasingFunction,
} from './commonTypes';
import type { DerivedValue as _DerivedValue } from './hook/useDerivedValue';
import type {
  TransformStyleTypes as _TransformStyleTypes,
  Adaptable as _Adaptable,
  AdaptTransforms as _AdaptTransforms,
  AnimateProps as _AnimateProps,
} from './helperTypes';

import type { AnimatedScrollViewProps as _AnimatedScrollViewProps } from './component/ScrollView';
import type { FlatListPropsWithLayout as _FlatListPropsWithLayout } from './component/FlatList';

export { createAnimatedComponent } from './createAnimatedComponent';
export { AnimatedText as Text } from './component/Text';
export { AnimatedView as View } from './component/View';
export { AnimatedScrollView as ScrollView } from './component/ScrollView';
export { AnimatedImage as Image } from './component/Image';
export { ReanimatedFlatList as FlatList } from './component/FlatList';
export {
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
} from './ConfigHelper';
/**
 * @deprecated Please import `Extrapolate` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type Extrapolate = typeof _Extrapolate;
/**
 * @deprecated Please import `SharedValue` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */

export type SharedValue<T> = _SharedValue<T>;
/**
 * @deprecated Please import `DerivedValue` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type DerivedValue<T> = _DerivedValue<T>;
/**
 * @deprecated Please import `Adaptable` directly from `react-native-reanimated`
 *   instead of `Animated` namespace.
 */
export type Adaptable<T> = _Adaptable<T>;
/**
 * @deprecated Please import `TransformStyleTypes` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type TransformStyleTypes = _TransformStyleTypes;
/**
 * @deprecated Please import `AdaptTransforms` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AdaptTransforms<T> = _AdaptTransforms<T>;
/**
 * @deprecated Please import `AnimatedTransform` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedTransform = _AnimatedTransform;
/**
 * @deprecated Please import `AnimateStyle` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimateStyle<S> = _AnimateStyle<S>;
/**
 * @deprecated Please import `StylesOrDefault` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type StylesOrDefault<S> = _StylesOrDefault<S>;
/**
 * @deprecated Please import `AnimateProps` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimateProps<P extends object> = _AnimateProps<P>;
/**
 * @deprecated Please import `EasingFunction` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type EasingFunction = _EasingFunction;
/**
 * @deprecated Please import `AnimatedScrollViewProps` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedScrollViewProps = _AnimatedScrollViewProps;
/**
 * @deprecated Please import `FlatListPropsWithLayout` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type FlatListPropsWithLayout<T> = _FlatListPropsWithLayout<T>;
