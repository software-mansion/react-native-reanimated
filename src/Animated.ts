import type { Extrapolate as _Extrapolate } from './reanimated2/interpolateColor';
import type { SharedValue as _SharedValue } from './reanimated2/commonTypes';
import type { DerivedValue as _DerivedValue } from './reanimated2/hook/useDerivedValue';
import type {
  TransformStyleTypes as _TransformStyleTypes,
  Adaptable as _Adaptable,
  AdaptTransforms as _AdaptTransforms,
  AnimatedTransform as _AnimatedTransform,
  AnimateStyle as _AnimateStyle,
  StylesOrDefault as _StylesOrDefault,
  AnimateProps as _AnimateProps,
} from './reanimated2/helperTypes';
import type { EasingFunction as _EasingFunction } from './reanimated2/Easing';
import {
  addWhitelistedNativeProps as _addWhitelistedNativeProps,
  addWhitelistedUIProps as _addWhitelistedUIProps,
} from './ConfigHelper';
import type { AnimatedScrollViewProps as _AnimatedScrollViewProps } from './reanimated2/component/ScrollView';
import type { FlatListPropsWithLayout as _FlatListPropsWithLayout } from './reanimated2/component/FlatList';
export { default as createAnimatedComponent } from './createAnimatedComponent';

export { AnimatedText as Text } from './reanimated2/component/Text';
export { AnimatedView as View } from './reanimated2/component/View';
export { AnimatedScrollView as ScrollView } from './reanimated2/component/ScrollView';
export { AnimatedImage as Image } from './reanimated2/component/Image';
export { ReanimatedFlatList as FlatList } from './reanimated2/component/FlatList';
/**
 * @deprecated Please import `Extrapolate` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type Extrapolate = typeof _Extrapolate;
/**
 * @deprecated Please import `SharedValue` directly from `react-native-reanimated` instead of `Animated` namespace.
 */

export type SharedValue<T> = _SharedValue<T>;
/**
 * @deprecated Please import `DerivedValue` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type DerivedValue<T> = _DerivedValue<T>;
/**
 * @deprecated Please import `Adaptable` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type Adaptable<T> = _Adaptable<T>;
/**
 * @deprecated Please import `TransformStyleTypes` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type TransformStyleTypes = _TransformStyleTypes;
/**
 * @deprecated Please import `AdaptTransforms` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AdaptTransforms<T> = _AdaptTransforms<T>;
/**
 * @deprecated Please import `AnimatedTransform` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedTransform = _AnimatedTransform;
/**
 * @deprecated Please import `AnimateStyle` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimateStyle<S> = _AnimateStyle<S>;
/**
 * @deprecated Please import `StylesOrDefault` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type StylesOrDefault<S> = _StylesOrDefault<S>;
/**
 * @deprecated Please import `AnimateProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimateProps<P extends object> = _AnimateProps<P>;
/**
 * @deprecated Please import `EasingFunction` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type EasingFunction = _EasingFunction;
/**
 * @deprecated Please import `addWhitelistedNativeProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export const addWhitelistedNativeProps = _addWhitelistedNativeProps;
/**
 * @deprecated Please import `addWhitelistedUIProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export const addWhitelistedUIProps = _addWhitelistedUIProps;
/**
 * @deprecated Please import `AnimatedScrollViewProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimatedScrollViewProps = _AnimatedScrollViewProps;
/**
 * @deprecated Please import `FlatListPropsWithLayout` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type FlatListPropsWithLayout<T> = _FlatListPropsWithLayout<T>;
