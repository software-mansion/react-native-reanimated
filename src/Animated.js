import { LogBox } from 'react-native';
import createAnimatedComponent from './createAnimatedComponent';
import {
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
} from './ConfigHelper';
import * as reanimated1 from './reanimated1';
import ReanimatedComponents from './reanimated2/component';

const Animated = {
  // components
  ...ReanimatedComponents,
  createAnimatedComponent,
  // configuration
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
  // reanimated 1
  ...reanimated1,
};

export * from './reanimated2';
export * from './reanimated1';
export default Animated;

// I think we can ignore this message as long as Gesture Handler doesn't
// try to load the Reanimated module. I figured it should be here instead of
// RNGH because this prevents the message from being displayed for all
// versions of RNGH.
LogBox.ignoreLogs([
  'RCTBridge required dispatch_sync to load RNGestureHandlerModule. This may lead to deadlocks',
]);
