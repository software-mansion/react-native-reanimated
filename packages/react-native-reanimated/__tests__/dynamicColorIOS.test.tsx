'use strict';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Pressable, Text } from 'react-native';

import Animated, {
  DynamicColorIOS,
  interpolateColor,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from '../src';

const AnimatedPressable = Animated.createAnimatedComponent(Pressable);

const LIGHT_COLORS = ['rgba(56, 172, 221, 1)', 'rgba(87, 180, 149, 1)'];
const DARK_COLORS = ['rgba(181, 141, 241, 1)', 'rgba(255, 98, 89, 1)'];

const AnimatedComponent = () => {
  const pressed = useSharedValue(0);

  const handlePress = () => {
    pressed.value = withTiming(pressed.value === 0 ? 1 : 0);
  };

  const animatedStyle = useAnimatedStyle(() => {
    const lightColor = interpolateColor(pressed.value, [0, 1], LIGHT_COLORS);
    const darkColor = interpolateColor(pressed.value, [0, 1], DARK_COLORS);

    return {
      width: withTiming(pressed.value === 0 ? 100 : 200),
      backgroundColor: DynamicColorIOS({
        light: lightColor,
        dark: darkColor,
      }),
    };
  });

  return (
    <AnimatedPressable
      testID={'pressable'}
      onPress={handlePress}
      style={[animatedStyle, { height: 100 }]}>
      <Text>Button</Text>
    </AnimatedPressable>
  );
};

const BackgroundParsedStyle = {
  backgroundColor: {
    dynamic: {
      light: LIGHT_COLORS[0],
      dark: DARK_COLORS[0],
      highContrastLight: undefined,
      highContrastDark: undefined,
    },
  },
};

const InterpolatedBackgroundParsedStyle = {
  backgroundColor: {
    dynamic: {
      light: LIGHT_COLORS[1],
      dark: DARK_COLORS[1],
      highContrastLight: undefined,
      highContrastDark: undefined,
    },
  },
};

const DefaultStyle = {
  width: 100,
  ...BackgroundParsedStyle,
};

describe('DynamicColorIOS test', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
  });

  test('colors are animating', () => {
    const style = DefaultStyle;
    const { getByTestId } = render(<AnimatedComponent />);
    const pressable = getByTestId('pressable');

    expect(pressable.props.style).toEqual([style, { height: 100 }]);

    fireEvent.press(pressable);
    jest.advanceTimersByTime(600);

    expect(pressable).toHaveAnimatedStyle(InterpolatedBackgroundParsedStyle);

    const rendered = render(<AnimatedComponent />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
  test('other props are not affected', () => {
    const style = DefaultStyle;
    const { getByTestId } = render(<AnimatedComponent />);
    const pressable = getByTestId('pressable');

    expect(pressable.props.style).toEqual([style, { height: 100 }]);

    fireEvent.press(pressable);
    jest.advanceTimersByTime(600);
    expect(pressable).toHaveAnimatedStyle({
      width: 200,
    });
  });
});
