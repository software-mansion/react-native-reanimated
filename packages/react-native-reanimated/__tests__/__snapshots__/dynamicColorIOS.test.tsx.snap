// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicColorIOS test colors are animating 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  jestAnimatedProps={
    {
      "value": {},
    }
  }
  jestAnimatedStyle={
    {
      "value": {
        "backgroundColor": {
          "dynamic": {
            "dark": "rgba(181, 141, 241, 1)",
            "highContrastDark": undefined,
            "highContrastLight": undefined,
            "light": "rgba(56, 172, 221, 1)",
          },
        },
        "width": 100,
      },
    }
  }
  jestInlineStyle={
    [
      {
        "height": 100,
      },
    ]
  }
  nativeID="1"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": {
          "dynamic": {
            "dark": "rgba(181, 141, 241, 1)",
            "highContrastDark": undefined,
            "highContrastLight": undefined,
            "light": "rgba(56, 172, 221, 1)",
          },
        },
        "width": 100,
      },
      {
        "height": 100,
      },
    ]
  }
  testID="pressable"
>
  <Text>
    Button
  </Text>
</View>
`;
